### ISSUE TITLE
OSMnx 2.0 Migration Guide

### ISSUE THREAD

**Issue by gboeing** [gboeing/osmnx#1123]
# Overview

OSMnx 2.0.0 was released in November 2024. This major release includes some breaking changes (including removing previously deprecated functionality) that are not backwards compatible with v1. This issue provides guidance on migrating from v1 to v2.

# Migrating to v2

If you want to ensure your code is fully compatible with OSMnx v2 before upgrading to it, you can install the latest v1 release (currently `1.9.4`), then run your code. The latest v1 release will remain compatible with the v1 API, but any deprecated functionality you're using will issue a FutureWarning with simple guidance to make it v2 compatible. Additionally, you can:

- refer to the new official OSMnx [reference paper](https://geoffboeing.com/publications/osmnx-paper/) up-to-date for v2
- check out the [Examples Gallery](https://github.com/gboeing/osmnx-examples), which has been updated for v2 compatibility if you're looking for sample code
- see the [changelog](https://github.com/gboeing/osmnx/blob/main/CHANGELOG.md) for details

# Key changes

The three biggest overarching changes in 2.0 are:

1. The package is now fully type-annotated throughout.
2. Several modules have been refactored or wholly rewritten for performance optimization.
3. Flexible and customizable graph simplification and node consolidation.
4. The package's API has been streamlined and made more consistent, including some breaking changes and some moved/renamed functions.

In addition to these, there have been many smaller enhancements, fixes, and changes. See the [changelog](https://github.com/gboeing/osmnx/blob/main/CHANGELOG.md) for details and see the latest [documentation](https://osmnx.readthedocs.io/en/latest/) for an up-to-date user reference.

**Comment by gboeing** [comment]
The first OSMnx v2.0.0 pre-release (beta0) has been released. Testers needed!

**Comment by EwoutH** [comment]
Briefly tested beta0 on a few of my existing notebooks. Not major issues so far!

**Comment by jGaboardi** [comment]
I'm also seeing no major/unexpected issues thus far.

**Comment by csebastiao** [comment]
No issues observed on a couple of my typical use cases !

Noting there in case the FutureWarning coming from Geopandas in the [1.0.0 latest version](https://github.com/geopandas/geopandas/issues/3201) when calling `geopandas.GeoSeries.unary_union()`, in `osmnx.graph.graph_from_place()`, or `osmnx.plot.plot_figure_ground()` for instance.

**Comment by gboeing** [comment]
Thanks @csebastiao. I'll make sure to resolve that FutureWarning before the next 2.0 beta release. I'm also planning to release a final patch version of 1.x at some point with pinned max dependency versions to avoid future dependency deprecation issues.

**Comment by QuentinMV** [comment]
Hi everyone.

I'm having some issues and i hope someone can help me. Im coding on 2 environnements ( databricks & visual studio code/jupyter ). I use osmx to create graphs to find the shortest path between 2 points. Everything works well on databricks but i have problems in Visual on this line : graph = ox.graph_from_bbox(north=north, south=south, east=east, west=west, network_type="all"). I have a furtur warning, which is not very important for now ( I suppose ), but this warning : 

UnboundLocalError: cannot access local variable 'response' where it is not associated with a value

I dont know where it comes from, because everything works well on databricks, i checked the proxy on visual, everything ok. There is more details in the error ( i put it below ) but i think the main issue is the one I put before. I dont know if my issue is connected to osmx but if you can help me, or just say the problem is not from osmx it will help me a lot !

Thanks ( sry for my english, not main language )

Full error :

[C:\Users\qmart261\AppData\Local\Temp\ipykernel_21756\1808055877.py:1](file:///C:/Users/qmart261/AppData/Local/Temp/ipykernel_21756/1808055877.py:1): FutureWarning: The `north`, `south`, `east`, and `west` parameters are deprecated and will be removed in the v2.0.0 release. Use the `bbox` parameter instead. See the OSMnx v2 migration guide: https://github.com/gboeing/osmnx/issues/1123
  graph = ox.graph_from_bbox(north=40, south=40.1, east=39.9, west=40.2, network_type="all")
---------------------------------------------------------------------------
SSLCertVerificationError                  Traceback (most recent call last)
File ~\AppData\Roaming\Python\Python312\site-packages\urllib3\connectionpool.py:467, in HTTPConnectionPool._make_request(self, conn, method, url, body, headers, retries, timeout, chunked, response_conn, preload_content, decode_content, enforce_content_length)
    [466](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/qmart261/OneDrive%20-%20Cafdoc/Documents/Visual/creche/~/AppData/Roaming/Python/Python312/site-packages/urllib3/connectionpool.py:466) try:
--> [467](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/qmart261/OneDrive%20-%20Cafdoc/Documents/Visual/creche/~/AppData/Roaming/Python/Python312/site-packages/urllib3/connectionpool.py:467)     self._validate_conn(conn)
    [468](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/qmart261/OneDrive%20-%20Cafdoc/Documents/Visual/creche/~/AppData/Roaming/Python/Python312/site-packages/urllib3/connectionpool.py:468) except (SocketTimeout, BaseSSLError) as e:

File ~\AppData\Roaming\Python\Python312\site-packages\urllib3\connectionpool.py:1099, in HTTPSConnectionPool._validate_conn(self, conn)
   [1098](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/qmart261/OneDrive%20-%20Cafdoc/Documents/Visual/creche/~/AppData/Roaming/Python/Python312/site-packages/urllib3/connectionpool.py:1098) if conn.is_closed:
-> [1099](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/qmart261/OneDrive%20-%20Cafdoc/Documents/Visual/creche/~/AppData/Roaming/Python/Python312/site-packages/urllib3/connectionpool.py:1099)     conn.connect()
   [1101](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/qmart261/OneDrive%20-%20Cafdoc/Documents/Visual/creche/~/AppData/Roaming/Python/Python312/site-packages/urllib3/connectionpool.py:1101) # TODO revise this, see https://github.com/urllib3/urllib3/issues/2791

File ~\AppData\Roaming\Python\Python312\site-packages\urllib3\connection.py:653, in HTTPSConnection.connect(self)
    [651](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/qmart261/OneDrive%20-%20Cafdoc/Documents/Visual/creche/~/AppData/Roaming/Python/Python312/site-packages/urllib3/connection.py:651) server_hostname_rm_dot = server_hostname.rstrip(".")
--> [653](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/qmart261/OneDrive%20-%20Cafdoc/Documents/Visual/creche/~/AppData/Roaming/Python/Python312/site-packages/urllib3/connection.py:653) sock_and_verified = _ssl_wrap_socket_and_match_hostname(
    [654](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/qmart261/OneDrive%20-%20Cafdoc/Documents/Visual/creche/~/AppData/Roaming/Python/Python312/site-packages/urllib3/connection.py:654)     sock=sock,
    [655](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/qmart261/OneDrive%20-%20Cafdoc/Documents/Visual/creche/~/AppData/Roaming/Python/Python312/site-packages/urllib3/connection.py:655)     cert_reqs=self.cert_reqs,
    [656](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/qmart261/OneDrive%20-%20Cafdoc/Documents/Visual/creche/~/AppData/Roaming/Python/Python312/site-packages/urllib3/connection.py:656)     ssl_version=self.ssl_version,
    [657](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/qmart261/OneDrive%20-%20Cafdoc/Documents/Visual/creche/~/AppData/Roaming/Python/Python312/site-packages/urllib3/connection.py:657)     ssl_minimum_version=self.ssl_minimum_version,
    [658](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/qmart261/OneDrive%20-%20Cafdoc/Documents/Visual/creche/~/AppData/Roaming/Python/Python312/site-packages/urllib3/connection.py:658)     ssl_maximum_version=self.ssl_maximum_version,
    [659](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/qmart261/OneDrive%20-%20Cafdoc/Documents/Visual/creche/~/AppData/Roaming/Python/Python312/site-packages/urllib3/connection.py:659)     ca_certs=self.ca_certs,
    [660](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/qmart261/OneDrive%20-%20Cafdoc/Documents/Visual/creche/~/AppData/Roaming/Python/Python312/site-packages/urllib3/connection.py:660)     ca_cert_dir=self.ca_cert_dir,
    [661](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/qmart261/OneDrive%20-%20Cafdoc/Documents/Visual/creche/~/AppData/Roaming/Python/Python312/site-packages/urllib3/connection.py:661)     ca_cert_data=self.ca_cert_data,
    [662](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/qmart261/OneDrive%20-%20Cafdoc/Documents/Visual/creche/~/AppData/Roaming/Python/Python312/site-packages/urllib3/connection.py:662)     cert_file=self.cert_file,
    [663](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/qmart261/OneDrive%20-%20Cafdoc/Documents/Visual/creche/~/AppData/Roaming/Python/Python312/site-packages/urllib3/connection.py:663)     key_file=self.key_file,
...
    [163](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/qmart261/OneDrive%20-%20Cafdoc/Documents/Visual/creche/~/AppData/Roaming/Python/Python312/site-packages/osmnx/_overpass.py:163)     return default_duration
    [164](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/qmart261/OneDrive%20-%20Cafdoc/Documents/Visual/creche/~/AppData/Roaming/Python/Python312/site-packages/osmnx/_overpass.py:164) except (AttributeError, IndexError, ValueError):  # pragma: no cover
    [165](https://file+.vscode-resource.vscode-cdn.net/c%3A/Users/qmart261/OneDrive%20-%20Cafdoc/Documents/Visual/creche/~/AppData/Roaming/Python/Python312/site-packages/osmnx/_overpass.py:165)     # cannot parse output, log error and return default duration

UnboundLocalError: cannot access local variable 'response' where it is not associated with a value

**Comment by EwoutH** [comment]
For me, the testing timeline doesn't need to be this long. I'm fine with two betas a few weeks apart and then one or two release candidates.

A new, stable version of osmnx before the academic year starts also has it's advantages.

Of course I understand there might be other interest or constraints at play, but I wanted to share my perspective.

**Comment by gboeing** [comment]
@EwoutH thanks. I'm specifically avoiding releasing breaking changes before the semester. Given OSMnx's use in academia, I have learned that *many* users won't even see the deprecation warnings from v1.9.x unless we give them a semester with it. Ideally their update cycle will expose them to those warnings this fall so they can update their code before 2.0 appears just after the semester.

Beta 1 should arrive within the next two weeks. And prior to the official release of 2.0, users can always pip install the beta to get the latest enhancements.

**Comment by EwoutH** [comment]
Can we add a _Breaking changes_ section somewhere (here or in the changelog)? To make things like changing the bounds order stand our a little more.

**Comment by EwoutH** [comment]
This code, which worked on OSMnx 1.x:
```Python
road_network = ox.load_graphml("merged_network.graphml")

city_name = "Rotterdam"
surrounding_area_name = "South Holland"
color_map = {city_name: 'red', surrounding_area_name: 'blue'}

edge_colors = [color_map[road_network.edges[edge]['network']] for edge in road_network.edges]
```
Now breaks on 2.0.
```
TypeError                                 Traceback (most recent call last)
Cell In[13], line 3
      1 # Edge colors
      2 color_map = {city_name: 'red', surrounding_area_name: 'blue'}
----> 3 edge_colors = [color_map[road_network.edges[edge]['network']] for edge in road_network.edges]
      5 # Generate a list of node populations to scale the node sizes for visualization
      6 populations = [int(data['population_per_node']) for _, data in road_network.nodes(data=True)]

TypeError: unhashable type: 'list'
```
Has anything changed in the edge view?

**Comment by EwoutH** [comment]
Sorry, it looks like this isn't a OSMnx, since the same error appear when using OSMnx 1.9.3.

Did anything change in GeoPandas, NetworkX or an other dependency?

**Comment by gboeing** [comment]
> Did anything change in GeoPandas, NetworkX or an other dependency?

I can't guess because I can't reproduce your error. However, if it is a dependency issue for OSMnx, it'll be resolved on the OSMnx side in the 1.9.4 release as we have pinned max dependency versions.

**Comment by EwoutH** [comment]
Yeah I think it was something on my side. If I encounter it again I will create a minimal reproducible example.

**Comment by Robinlovelace** [comment]
As per the above commit, we're trying to install a version of `osmnx` that is compatible with the latest version of `geopandas` but to no avail currently.

Error message:

```
The conflict is caused by:
    The user requested geopandas>=1.0.0-alpha1
    osmnx 1.9.4 depends on geopandas<0.15 and >=0.12

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict
```

As far as I can tell from the conversation above, it should be compatible with geopandas 1.0.0 but will check... Any ideas welcome, see here for details: https://github.com/geocompx/geocompy/issues/241

**Comment by Robinlovelace** [comment]
OK, seems it's an issue with installing the latest version of `osmnx`

https://github.com/gboeing/osmnx/blob/8df6d7ab8b92bb7dac70822ea621e53d2d529f65/osmnx/_version.py#L3

**Comment by Robinlovelace** [comment]
Have tried to install the latest version with 

```
pip install --pre osmnx
```

but that seems to get v1.9.4, right?

**Comment by Robinlovelace** [comment]
> but that seems to get v1.9.4, right?

Update: no, after testing with the following reproducible example:

```
docker run -it python /bin/bash
root@77d9f959ad9d:/# pip install --pre osmnx
root@77d9f959ad9d:/# python
>>> import osmnx
>>> print(osmnx.__version__)
2.0.0b1
```

**Comment by Robinlovelace** [comment]
Will try with:

```bash
pip install --no-cache-dir --upgrade --pre osmnx
```

**Comment by Robinlovelace** [comment]
Source: https://stackoverflow.com/a/44008863

**Comment by gboeing** [comment]
I would expect https://github.com/gboeing/osmnx/issues/1123#issuecomment-2271975575 to work.

**Comment by Robinlovelace** [comment]
Yes it did, although that was being overridden in our workflow. Fixed longer term (in a way that will keep updating to the latest versions) with the following command before running `pip install -U --upgrade-strategy eager -r requirements.txt`, sharing below in case of use/interest to others:

```bash
pur -r requirements.txt --pre osmnx
```

Source: https://github.com/geocompx/geocompy/commit/30ffd68119c47df27c59780c0172804d905500e1

**Comment by Robinlovelace** [comment]
By the way, as a relative Python novice, I'm interested in best practice and would welcome any suggestions on how to improve our set-up/installation/CI game.

**Comment by jamongadejoa28** [comment]
Numpy's latest version of 2.0.0rc1 and osmnx's 2.0.0b1 are incompatible. When converting osm data, when I look at the log history, when the tool invokes the coordinate values, it prepends a number with float64. If you look at the 2.0.0 version release of numpy, you can see that float_ has been deleted and replaced by float64. Please let me know the most stable versions of numpy and other tools that are compatible.

**Comment by gboeing** [comment]
@jamongadejoa28 OSMnx 2.0 should be fully compatible with Numpy 2.0. If you discover a specific incompatibility, please open a new issue with complete troubleshooting details so we can resolve it.

**Comment by EwoutH** [comment]
I'm now changing the order of the coordinates in public examples. It's a bit weird to need to stay on a beta release for almost half a year. Maybe I'm too early, but I would support going a lot faster with the 2.0 release schedule.

Even huge projects as NumPy or Pandas have way shorter pre-release periods.

Maybe something like:

- 2.0.0b2 now
- 2.0.0rc0 september 1st
- 2.0.0 september 15th

**Comment by jakirkham** [comment]
Am I understanding correctly that OSMnx 2.0 will be the first release with NumPy 2 compatibility (and the OSMnx 1.x series will be limited to NumPy 1)?

Sorry if I'm asking something that is clarified elsewhere, had a hard time finding the answer when searching (with the exception of `pyproject.toml` constraints)

**Comment by gboeing** [comment]
@jakirkham correct. We're not updating the 1.x branch further (outside of bug fixes or deprecation warnings necessary for transitioning to v2). However, if dependency version incompatibility poses a real problem for the community in the meantime, I'm always happy to review a new issue documenting the problem and proposing a minimal PR to resolve it.

**Comment by gboeing** [comment]
Beta 2 has been released with:
  - support for `custom_filter` union queries in the `graph` module
  - faster, optimized geometry construction in the `features` module
  - removing the optional `gdal` dependency and replacing it with `rio-vrt`
  - new `buffer_geometry` helper function (a small user quality-of-life improvement)

**Comment by gboeing** [comment]
OSMnx 2.0.0 has been released.
